url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/newModel4d/TimeStepBuilder.java

```diff
@@ -0,0 +1,436 @@
+			Map<Integer, Point3D> newPoints, Date date) {
+
+		// check if it is the first timestep
+		if (component.getTimesteps().isEmpty()) {
+
+			firstStep(component, newPoints, date);
+
+			// if there are already some timesteps in the Map, check if the new
+			// timestep is higher than the last one
+
+		} else if (!date.before(component.getTimesteps().getLast())) {
+
+			// Now we need to know if the user already added the geometry for
+			// the last Postobject (can be whether the first object or the
+			// Postobject of a new interval)
+
+			// We do not need this anymore! One component is always in one time
+			// interval
+			// if (geometry.size() < countToplogogyChanges(component))
+			// throw new IllegalArgumentException(
+			// "The net topology changed. You need to add the geometrydata for the last Postobject first. Call the addGeometry() function.");
+
+			// TODO: Das muss irgendwie raus... P+R wird jetzt durch die
+			// Komponenten geregelt!
+
+			// this is the Polthier und Rumpf model
+			// do we have a change of topology?
+			// check if this is an Post-object!
+			if (component.getTimesteps().getLast().equals(date)) {
+
+				// TODO Wird hier nicht funktionieren, da wir neue Componenten
+				// haben fuer die P+R Geschichte!
+				// Also: Dead Code!!! Wahrscheinlich
+				DB3DLogger.logger
+						.log(Level.INFO,
+								"Polthier und Rumpf! Should not be called (TimeStepBuilder.java)");
+
+				polthierAndRumpf(component, newPoints, date);
+
+				// Notify corresponding net
+				// TODO: vorher oder hier?
+				// Bei dem GOCAD Importer wird das quasi auch gemacht! Checken!
+
+				
+				// no Post-object, should be the same topology
+			} else {
+
+				addNormalTimeStep(component, newPoints, date);
+			}
+		}
+	}
+
+	/**
+	 * Adds a new timestep to the component object. This method checks if the
+	 * new data has the same topology as the last timestep. Checks the
+	 * correlation of partitions between the current and the last timestep.
+	 * 
+	 * @param newPoints
+	 * @param date
+	 */
+	private static void addNormalTimeStep(Component4D component,
+			Map<Integer, Point3D> newPoints, Date date) {
+
+		// check if the new step has the same ids of points as the last
+		// step:
+		Iterator<Integer> it = newPoints.keySet().iterator();
+
+		Map<Integer, List<Point3D>> pointTubes = component.getPointTubes();
+
+		LinkedList<Date> timesteps = component.getTimesteps();
+
+		while (it.hasNext()) {
+			if (pointTubes.get(it.next()).get(timesteps.size() - 1) == null)
+				throw new IllegalArgumentException(
+						"New Object is neither a Postobject nor it fits the size of the last object");
+
+		}
+
+		// are there more points at the last timestep than in the new
+		// one?
+
+		// TODO: entrys nehmen oder einfach irgendwelche groessen der container
+		// klassen!
+		Set<Integer> ids = pointTubes.keySet();
+		int cnt = 0;
+		for (final Integer id : ids) {
+			if (pointTubes.get(id).get(timesteps.size() - 1) != null)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.